### PR TITLE
Mir Audit: Update availability module and have orderer verify certificates

### LINKED ISSUE #335: Orderer's preprepare not verifying certificate
The current Orderer's treatment of a preprepare msg sent by the primary does not verify the certificate. A malicious proposer could propose a bogus certificate and have it decided.

### PR DESCRIPTION
**Pull request by ranchalp**
Fixes #335 , 
Also applies fix to potential bug #340.
Implements #343 
Hardcoded a limit of certificates to precompute and provide per request which can be set by the application later on. (at the moment set by ISS to limit=5)

### CODE REVIEW

**Comment by ranchalp** [comment]
Did make format and make lint locally but workflow fails for some reason. I would not worry.

**Comment by ranchalp** [comment]
@matejpavlovic when reviewing please also have a look at #340 and the fix. I had the issue when using string's widely (as I had to choose one of the two, either string or []byte, when using the DSL codegen) for the reason explained in the bug, but I still think it is odd that this problem did not arise in previous tests/deployments.

**Comment by ranchalp** [comment]
One last note, I do not know why (perhaps merging upon rebase) but I had to reset file `leaderselectionpolicy.go`. It has been resetted to existing HEAD of master (except for removed unused private functions to make lint happy), so no changes should be visible upon squash.

File: pkg/availability/multisigcollector/internal/parts/certverification/certverification.go [hunk 1]
```diff
@@ -4,103 +4,179 @@ import (
...
-		dsl.VerifyNodeSigs(m, mc.Crypto,
-			/*data*/ sliceutil.Repeat(sigMsg, len(mscCert.Signers)),
-			/*signatures*/ mscCert.Signatures,
-			/*nodeIDs*/ t.NodeIDSlice(mscCert.Signers),
-			/*context*/ &verifySigsInCertContext{origin},
-		)
+		state.RequestState[reqID] = &RequestState{
+			certVerifiedValid: make(map[t.BatchIDString]bool),
+		}
+		nonEmptyCerts := make([]*mscpbtypes.Cert, 0)
```
Comment (matejpavlovic) [inline pkg/availability/multisigcollector/internal/parts/certverification/certverification.go:61]: Do we need to use this and iterate twice? Can we directly issue the signature verification request if the cert is not empty? (And just have a flag for the case that all certificates are empty.)
I think it would simplify the code.

File: pkg/availability/multisigcollector/internal/parts/certverification/certverification.go [hunk 2]
```diff
@@ -4,103 +4,179 @@ import (
...
 
 	// When the signatures in a certificate are verified, output the result of certificate verification.
 	dsl.UponNodeSigsVerified(m, func(nodeIDs []t.NodeID, errs []error, allOK bool, context *verifySigsInCertContext) error {
+		reqID := context.reqID
+
+		if _, ok := state.RequestState[reqID]; !ok {
+			return nil
+		}
+
+		state.RequestState[reqID].certVerifiedValid[t.BatchIDString(string(context.cert.BatchId))] = allOK
```
Comment (matejpavlovic) [inline pkg/availability/multisigcollector/internal/parts/certverification/certverification.go:96]: The string cast seems to be superfluous here (at least my IDE says so).

File: pkg/availability/multisigcollector/internal/parts/certverification/certverification.go [hunk 3]
```diff
@@ -4,103 +4,179 @@ import (
...
+		// if we get here, that means so far all received signatures for all certificates in the reqID are verified valid or yet to be verified
+		allVerifiedValid := true
+		for _, certVerifiedValid := range state.RequestState[reqID].certVerifiedValid {
+			if !certVerifiedValid {
+				allVerifiedValid = false
+				break
+			}
+		}
+
+		if _, ok := state.RequestState[reqID]; allVerifiedValid && ok {
```
Comment (matejpavlovic) [inline pkg/availability/multisigcollector/internal/parts/certverification/certverification.go:114]: Haven't we already checked (on line 92) that the request is still pending?

File: pkg/availability/multisigcollector/multisigcollector.go [hunk 4]
```diff
@@ -66,10 +65,12 @@ func NewReconfigurableModule(mc *ModuleConfig, nodeID t.NodeID, logger logging.L
 			func(mscID t.ModuleID, params *factorymodulepb.GeneratorParams) (modules.PassiveModule, error) {
 
 				// Extract the IDs of the nodes in the membership associated with this instance
-				m := params.Type.(*factorymodulepb.GeneratorParams_MultisigCollector).MultisigCollector.Membership
+				mscParams := params.Type.(*factorymodulepb.GeneratorParams_MultisigCollector).MultisigCollector
+				m := mscParams.Membership
```
Comment (matejpavlovic) [inline pkg/availability/multisigcollector/multisigcollector.go:69]: `m` is only used in the next line. Maybe we can get rid of that variable and use `mscParams.Membership` directly?

File: protos/availabilitypb/mscpb/mscpb.proto [hunk 5]
```diff
@@ -54,18 +70,26 @@ message ProvideBatchMessage {
...
+}
+
 // ============================================================
 // Instantiation parameters
 // ============================================================
 
 message InstanceParams {
   commonpb.Membership membership = 1;
+  uint64 limit = 2;
+  uint64 segment_length = 3;
```
Comment (matejpavlovic) [inline protos/availabilitypb/mscpb/mscpb.proto:94]: let's rename this to `max_requests` or something along those lines, not related to segments.

File: pkg/orderers/pbftgoodcase.go [hunk 6]
```diff
@@ -86,7 +86,7 @@ func (orderer *Orderer) requestNewCert() *events.EventList {
 
 	// Emit the CertRequest event.
 	// Operation continues on reception of the CertReady event.
-	return orderer.requestCertOrigin()
+	return orderer.requestCertOrigin() //.PushBack(availabilityevents.ComputeCert(orderer.moduleConfig.Ava))
```
Comment (matejpavlovic) [inline pkg/orderers/pbftgoodcase.go:89]: Is this commented code still relevant?

File: pkg/orderers/pbftgoodcase.go [hunk 7]
```diff
@@ -119,8 +119,9 @@ func (orderer *Orderer) applyCertReady(cert *availabilitypb.Cert) (*events.Event
 // propose assumes that the state of the PBFT Orderer allows sending a new proposal
 // and does not perform any checks in this regard.
 func (orderer *Orderer) propose(cert *availabilitypb.Cert) (*events.EventList, error) {
+	//TODO Alejandro: argument must be slice of pointers to certs
```
Comment (matejpavlovic) [inline pkg/orderers/pbftgoodcase.go:122]: I don't quite understand... The argument still seems to be a single certificate.

File: pkg/orderers/pbftgoodcase.go [hunk 8]
```diff
@@ -190,6 +191,36 @@ func (orderer *Orderer) applyMsgPreprepare(preprepare *ordererspbftpb.Preprepare
...
+	context := orderer.notVerifiedPrepepareContext[0]
+	orderer.notVerifiedPrepepareContext = orderer.notVerifiedPrepepareContext[1:]
+
+	if !verified.Valid {
+		return events.EmptyList(), fmt.Errorf(verified.Err)
+	}
+	return orderer.finishUpPreprepare(context.prepepare, context.slot), nil
+}
+
+func (orderer *Orderer) finishUpPreprepare(preprepare *ordererspbftpb.Preprepare, slot *pbftSlot) *events.EventList {
```
Comment (matejpavlovic) [inline pkg/orderers/pbftgoodcase.go:223]: In fact, I think that in our particular case it is better to only check the validity of the certificates after agreeing on them. It is better for ISS finish the agreement on a certificate, even if it's invalid, and discard it later, then to wait for a timeout.

I suggest removing this check and addressing it in a separate PR.

Another reason is that it will make merging easier, as this code conflicts with the changes I made to implement the agreement on checkpoint certificates.

File: pkg/types/types.go [hunk 9]
```diff
@@ -93,70 +95,66 @@ func NodeIDSlicePb(nids []NodeID) []string {
...
+//
+//// Pb converts a TxID to its underlying native type.
+//func (id TxID) Pb() []byte {
+//	return []byte(id)
+//}
+//
+//// Bytes serializes the object to a sequence of bytes.
+//func (id TxID) Bytes() []byte {
+//	return []byte(id)
+//}
```
Comment (matejpavlovic) [inline pkg/types/types.go:109]: Commented code better to be removed.

File: pkg/types/types.go [hunk 10]
```diff
@@ -93,70 +95,66 @@ func NodeIDSlicePb(nids []NodeID) []string {
...
+//
+//// Pb converts a BatchID to its underlying native type.
+//func (id BatchID) Pb() []byte {
+//	return []byte(id)
+//}
+//
+//// Bytes serializes the object to a sequence of bytes.
+//func (id BatchID) Bytes() []byte {
+//	return []byte(id)
+//}
```
Comment (matejpavlovic) [inline pkg/types/types.go:143]: Remove commented code.

**Review by matejpavlovic — CHANGES_REQUESTED** [review]
Great job! Mostly minor comments.

As is, this PR has conflicts with the main branch (due to the consistent checkpointing update), but let's care about that later. When the PR is polished, I'll take care of the rebasing.

File: pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go [hunk 11]
```diff
@@ -43,63 +47,93 @@ func IncludeBatchReconstruction(
...
+
+		reqID := state.NextReqID
+		state.NextReqID++
+		state.RequestState[reqID] = &RequestState{decidedOrder: make([]t.BatchIDString, 0), Txs: make(map[t.BatchIDString][]*requestpbtypes.Request), ReqOrigin: origin}
+
+		mscCertsWrapper, ok := cert.Type.(*apbtypes.Cert_Mscs)
+		if !ok {
+			return fmt.Errorf("unexpected certificate type")
+		}
+		cs := mscCertsWrapper.Mscs
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go:62]: changed to `certs`

File: pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go [hunk 12]
```diff
@@ -43,63 +47,93 @@ func IncludeBatchReconstruction(
...
+			apbdsl.ProvideTransactions(m, origin.Module, []*requestpbtypes.Request{}, origin)
+			delete(state.RequestState, reqID)
+		}
+
 		return nil
 	})
 
 	// If the batch is present in the local storage, return it. Otherwise, ask the nodes that signed the certificate.
-	batchdbdsl.UponLookupBatchResponse(m, func(found bool, txs []*requestpb.Request, metadata []byte, context *lookupBatchLocallyContext) error {
+	batchdbpbdsl.UponLookupBatchResponse(m, func(found bool, txs []*requestpbtypes.Request, metadata []byte, context *lookupBatchLocallyContext) error {
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go:89]: I agree but it was there from before my edits on the module so I did not want to remove it yet. Removed now.

File: pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go [hunk 13]
```diff
@@ -1,33 +1,37 @@
...
 
-// RequestID is used to uniquely identify an outgoing request.
-type RequestID = uint64
-
 // RequestState represents the state related to a request on the source node of the request.
 // The node disposes of this state as soon as the request is completed.
 type RequestState struct {
-	BatchID   t.BatchID
-	ReqOrigin *apb.RequestTransactionsOrigin
+	decidedOrder []t.BatchIDString
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go:32]: done

File: pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go [hunk 14]
```diff
@@ -43,63 +47,93 @@ func IncludeBatchReconstruction(
...
 	// When receive a request for transactions, first check the local storage.
-	mscdsl.UponRequestTransactions(m, func(cert *mscpb.Cert, origin *apb.RequestTransactionsOrigin) error {
+	apbdsl.UponRequestTransactions(m, func(cert *apbtypes.Cert, origin *apbtypes.RequestTransactionsOrigin) error {
+		//TODO this switch might be removable with some refactoring
 		// NOTE: it is assumed that cert is valid.
-		batchdbdsl.LookupBatch(m, mc.BatchDB, t.BatchID(cert.BatchId), &lookupBatchLocallyContext{cert, origin})
+
+		reqID := state.NextReqID
+		state.NextReqID++
+		state.RequestState[reqID] = &RequestState{decidedOrder: make([]t.BatchIDString, 0), Txs: make(map[t.BatchIDString][]*requestpbtypes.Request), ReqOrigin: origin}
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go:56]: done

File: pkg/availability/multisigcollector/internal/common/common.go [hunk 15]
```diff
@@ -25,12 +26,15 @@ type ModuleConfig struct {
...
 type ModuleParams struct {
-	InstanceUID []byte     // unique identifier for this instance of BCB, used to prevent cross-instance replay attacks
-	AllNodes    []t.NodeID // the list of participating nodes
-	F           int        // the maximum number of failures tolerated. Must be less than (len(AllNodes)-1) / 2
+	InstanceUID   []byte             // unique identifier for this instance of BCB, used to prevent cross-instance replay attacks
+	AllNodes      []t.NodeID         // the list of participating nodes
+	F             int                // the maximum number of failures tolerated. Must be less than (len(AllNodes)-1) / 2
+	Certs         []*mscpbtypes.Cert // the list of generated certificates
+	Limit         int                // the maximum number of certificates to generate before a request is completed
+	SegmentLength int                // the number of sequence numbers per segment
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/common/common.go:34]: done

File: pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go [hunk 16]
```diff
@@ -43,63 +47,93 @@ func IncludeBatchReconstruction(
...
-			adsl.ProvideTransactions(m, t.ModuleID(context.origin.Module), txs, context.origin)
-			return nil
+			if _, ok := state.RequestState[context.reqID]; !ok {
+				return nil
+			}
+			saveAndFinish(m, context.reqID, txs, context.cert.BatchId, context.origin, &state)
+
+		} else {
+			eventpbdsl.SendMessage(m, mc.Net,
+				mscpbmsgs.RequestBatchMessage(mc.Self, context.cert.BatchId, context.reqID),
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go:98]: Done. Is that expected behavior from the network for *all* messages?

File: pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go [hunk 17]
```diff
@@ -43,63 +47,93 @@ func IncludeBatchReconstruction(
...
 
 	// If the batch is present in the local storage, return it. Otherwise, ask the nodes that signed the certificate.
-	batchdbdsl.UponLookupBatchResponse(m, func(found bool, txs []*requestpb.Request, metadata []byte, context *lookupBatchLocallyContext) error {
+	batchdbpbdsl.UponLookupBatchResponse(m, func(found bool, txs []*requestpbtypes.Request, metadata []byte, context *lookupBatchLocallyContext) error {
 		if found {
-			adsl.ProvideTransactions(m, t.ModuleID(context.origin.Module), txs, context.origin)
-			return nil
+			if _, ok := state.RequestState[context.reqID]; !ok {
+				return nil
+			}
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go:93]: Yes, done

File: pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go [hunk 18]
```diff
@@ -111,20 +145,18 @@ func IncludeBatchReconstruction(
 			return nil
 		}
 
-		if batchID != requestState.BatchID {
+		if !reflect.DeepEqual(batchID, context.batchID) {
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go:148]: At the moment `BatchID` is an unnamed type `BatchID = []byte`, meaning that no methods can be added to the type. Changing that would require tagging the field type in protobuffers, but a codegen bug is related to this issue #356.

I did get rid of reflection by just having `string(batchID) != string(context.batchID)` though.

File: pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go [hunk 19]
```diff
@@ -26,11 +31,13 @@ type RequestID = uint64
 // RequestState represents the state related to a request on the source node of the request.
 // The node disposes of this state as soon as the request is completed.
 type RequestState struct {
-	ReqOrigin *apb.RequestCertOrigin
-	BatchID   t.BatchID
+	ReqOrigin *apbtypes.RequestCertOrigin
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go:34]: It conceptually makes sense. At the moment the struct suffices but see previous code for an example of where this struct requires more than just the `origin` field.

File: pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go [hunk 20]
```diff
@@ -1,23 +1,28 @@
...
 	"github.com/filecoin-project/mir/pkg/util/maputil"
 )
 
 // State represents the state related to this part of the module.
 type State struct {
-	NextReqID    RequestID
-	RequestState map[RequestID]*RequestState
+	NextReqID      RequestID
+	RequestState   []*RequestState
+	Certificate    map[RequestID]*Certificate
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go:24]: Done

File: pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go [hunk 21]
```diff
@@ -149,25 +180,77 @@ func IncludeCreatingCertificates(
...
+
+	if len(certFinished) > 0 { // if any certificate is ready
+		maputil.FindAndDeleteAll(state.Certificate,
+			func(key RequestID, val *Certificate) bool {
+				_, ok := certsToDelete[t.BatchIDString(val.BatchID)]
+				return ok
+			}) // prevent duplicates
+
+		// prepare finished certs for proposal
+		_proposal := make([]*mscpbtypes.Cert, 0, len(certFinished))
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go:220]: done.

File: pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go [hunk 22]
```diff
@@ -149,25 +180,77 @@ func IncludeCreatingCertificates(
...
+		for _, _cert := range certFinished { // at least one
+			certNodes, certSigs := maputil.GetKeysAndValues(_cert.sigs)
+			cert := &mscpbtypes.Cert{BatchId: _cert.BatchID, Signers: certNodes, Signatures: certSigs}
+			_proposal = append(_proposal, cert)
+		}
+
+		proposal := &apbtypes.Cert{Type: &apbtypes.Cert_Mscs{Mscs: &mscpbtypes.Certs{
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go:227]: There is not because these are just protobufs structs and not actual events. The actual event starts with .NewCert

File: pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go [hunk 23]
```diff
@@ -149,25 +180,77 @@ func IncludeCreatingCertificates(
...
+			apbdsl.NewCert(m, requestingModule, proposal, requestState.ReqOrigin)
+			state.RemainingSeqNr--
+		}
+		// Dispose of the state associated with handled requests.
+		state.RequestState = make([]*RequestState, 0)
+
+	} else if sendEmptyIfNoneReady { // if none are ready and sendEmptyIfNoneReady is true, send empty cert
+		emptyProposal := &apbtypes.Cert{Type: &apbtypes.Cert_Mscs{Mscs: &mscpbtypes.Certs{
+			Certs: []*mscpbtypes.Cert{emptycert.EmptyCert()},
+		}}}
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go:243]: ditto

File: pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go [hunk 24]
```diff
@@ -149,25 +180,77 @@ func IncludeCreatingCertificates(
...
+			func(key RequestID, val *Certificate) bool {
+				_, ok := certsToDelete[t.BatchIDString(val.BatchID)]
+				return ok
+			}) // prevent duplicates
+
+		// prepare finished certs for proposal
+		_proposal := make([]*mscpbtypes.Cert, 0, len(certFinished))
+		for _, _cert := range certFinished { // at least one
+			certNodes, certSigs := maputil.GetKeysAndValues(_cert.sigs)
+			cert := &mscpbtypes.Cert{BatchId: _cert.BatchID, Signers: certNodes, Signatures: certSigs}
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go:223]: ditto

File: pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go [hunk 25]
```diff
@@ -101,32 +138,26 @@ func IncludeCreatingCertificates(
...
+		certificate.sigs[nodeID] = context.signature
 
-	// When F+1 signatures are collected, create and output a certificate.
-	dsl.UponCondition(m, func() error {
-		// Iterate over active outgoing requests.
-		// Most of the time, there is expected to be at most one active outgoing request.
-		for reqID, requestState := range state.RequestState {
-			if len(requestState.sigs) >= params.F+1 {
-				certNodes, certSigs := maputil.GetKeysAndValues(requestState.sigs)
+		newDue := len(certificate.sigs) >= params.F+1 // keep this here...
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go:149]: No, but the resulting bool is used after the line that changes the state. I cannot know if a new one is due after the call l to `sendIfReady` because if the certificates is ready to be sent then it deletes the certificate (and thus the check is not valid).

File: pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go [hunk 26]
```diff
@@ -149,25 +180,77 @@ func IncludeCreatingCertificates(
 
 	// When the id of the batch is computed, store the batch persistently.
 	mempooldsl.UponBatchIDResponse(m, func(batchID t.BatchID, context *computeIDOfReceivedBatchContext) error {
-		batchdbdsl.StoreBatch(m, mc.BatchDB, batchID, context.txIDs, context.txs, nil, /*metadata*/
+		batchdbpbdsl.StoreBatch(m, mc.BatchDB, batchID, context.txIDs, context.txs, nil, /*metadata*/
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go:183]: In that case, I would say the optimization is not even getting a reply from the batchdb. (removing the event sent back from batchdb). But noted.

File: pkg/availability/multisigcollector/internal/parts/certverification/certverification.go [hunk 27]
```diff
@@ -4,103 +4,179 @@ import (
...
+type RequestState struct {
+	certVerifiedValid map[t.BatchIDString]bool
+}
+
 // IncludeVerificationOfCertificates registers event handlers for processing availabilitypb.VerifyCert events.
 func IncludeVerificationOfCertificates(
 	m dsl.Module,
 	mc *common.ModuleConfig,
 	params *common.ModuleParams,
 	nodeID t.NodeID,
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certverification/certverification.go:38]: done.

File: pkg/availability/multisigcollector/internal/parts/certverification/certverification.go [hunk 28]
```diff
@@ -4,103 +4,179 @@ import (
...
-			return nil, fmt.Errorf("unknown node id: %v", t.NodeID(idRaw))
+		// Check that the identities of the source node and the signing nodes are valid.
+		allNodes := make(map[t.NodeID]struct{})
+		for _, id := range params.AllNodes {
+			allNodes[id] = struct{}{}
 		}
-	}
 
-	return mscCert, nil
+		for _, idRaw := range mscCert.Signers {
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certverification/certverification.go:167]: correct

File: pkg/availability/multisigcollector/internal/parts/certverification/certverification.go [hunk 29]
```diff
@@ -4,103 +4,179 @@ import (
...
 
-	// Check that the identities of the source node and the signing nodes are valid.
-	allNodes := make(map[t.NodeID]struct{})
-	for _, id := range params.AllNodes {
-		allNodes[id] = struct{}{}
-	}
+		// Check that the identities of the signing nodes are not repeated.
+		alreadySeen := make(map[t.NodeID]struct{})
+		for _, idRaw := range mscCert.Signers {
+			id := idRaw
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certverification/certverification.go:154]: correct.

File: pkg/availability/multisigcollector/internal/parts/certverification/certverification.go (hunk 1 above)
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certverification/certverification.go:61]: done

File: pkg/availability/multisigcollector/internal/parts/certverification/certverification.go (hunk 3 above)
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certverification/certverification.go:114]: correct

File: pkg/availability/multisigcollector/multisigcollector.go (hunk 4 above)
Comment (ranchalp) [reply pkg/availability/multisigcollector/multisigcollector.go:69]: done

File: protos/availabilitypb/mscpb/mscpb.proto (hunk 5 above)
Comment (ranchalp) [reply protos/availabilitypb/mscpb/mscpb.proto:94]: done

File: pkg/iss/iss.go [hunk 30]
```diff
@@ -621,25 +623,38 @@ func (iss *ISS) startEpoch(epochNr t.EpochNr) *events.EventList {
...
+	eventsOut.PushBackSlice(iss.initAvailability())
 
 	// Initialize the orderer modules for the current epoch.
 	eventsOut.PushBackList(iss.initOrderers())
-
+	//LEFT TO DO:
+	// 1- Make it pass tests
+	// 2- Verify that batchfetcher does the proper treatment to these batches of certs
+	// 3- Verify that ISS is calling the computecert at the availability module at the right place
+	// 4- Comment modified and new code (specially availability/parts/*)
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/iss/iss.go:634]: Sorry, missed that comment.

File: pkg/iss/iss.go [hunk 31]
```diff
@@ -621,25 +623,38 @@ func (iss *ISS) startEpoch(epochNr t.EpochNr) *events.EventList {
...
+	// bonus1 - Optimize?
 	return eventsOut
 }
 
 // initAvailability emits an event for the availability module to create a new submodule
 // corresponding to the current ISS epoch.
-func (iss *ISS) initAvailability() *eventpb.Event {
-	return factoryevents.NewModule(
+func (iss *ISS) initAvailability() []*eventpb.Event {
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/iss/iss.go:641]: done

File: pkg/orderers/pbftgoodcase.go (hunk 6 above)
Comment (ranchalp) [reply pkg/orderers/pbftgoodcase.go:89]: no. removed.

File: pkg/orderers/pbftgoodcase.go (hunk 7 above)
Comment (ranchalp) [reply pkg/orderers/pbftgoodcase.go:122]: The proto message `availabilitypb.proto:Cert` is a message class that contains a `oneof` field. Inside the `oneof` is where one can see that the proto message `Cert` is actually a `mscpb.Certs`. I decided not to change that abstraction of the `availabilitypb.proto:Cert` (for ISS, it does not matter what form, number or type a cert takes, it just needs to receive one from the availability module).

File: pkg/orderers/pbftgoodcase.go (hunk 8 above)
Comment (ranchalp) [reply pkg/orderers/pbftgoodcase.go:223]: Done, left comment to be addressed by future PR (once merge occurs). Happy to address it then.

File: pkg/orderers/protobufs.go [hunk 32]
```diff
@@ -26,17 +34,60 @@ func OrdererEvent(
...
+//							Module: orderer.moduleConfig.Self,
+//							Type: &apbtypes.RequestCertOrigin_ContextStore{ContextStore: &contextstorepbtypes.Origin{
+//								ItemID: 0, // TODO remove this parameter. It is deprecated as now ModuleID is a particular PBFT orderer.
+//							}},
+//						},
+//					}.Pb()}}})
+//}
+
+// Request availability module to verify the certificate from the preprepare message
+func (orderer *Orderer) verifyCert(preprepare *ordererspbftpb.Preprepare) *events.EventList {
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/orderers/protobufs.go:65]: To be noted for future PR.

File: pkg/types/types.go (hunk 9 above)
Comment (ranchalp) [reply pkg/types/types.go:109]: done.

File: pkg/types/types.go (hunk 10 above)
Comment (ranchalp) [reply pkg/types/types.go:143]: done

File: pkg/availability/multisigcollector/internal/parts/certverification/certverification.go (hunk 2 above)
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certverification/certverification.go:96]: done

**Comment by ranchalp** [comment]
Addressed comments. Left rebasing TODO as requested.

File: pkg/availability/multisigcollector/internal/common/common.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (matejpavlovic) [reply pkg/availability/multisigcollector/internal/common/common.go]: Can you update the description in the comment too?

File: pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (matejpavlovic) [reply pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go:103]: Yes, any network message can get lost arbitrary many times, or have an arbitrary delay.

File: pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (matejpavlovic) [reply pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go]: Yes I guess that's also better than the `reflect`.

But as far as I understand, `BatchID` is a named type defined in `types.go`
```go
// BatchID is a unique identifier of a batch.
type BatchID = []byte
```
You can define methods on this type the same way you could do for `struct` types (have a look at `NodeID` or `SeqNr` in `types.go`, for example).

Not 100% sure how it goes if you needed to have a `[]uint8` treated as `BatchID`, given the above definition, even though `byte` is just an alias for `uint8`...

Bottom line: Do whatever you prefer and resolve this convo ;-)

File: pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (matejpavlovic) [reply pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go]: If I remember it right, the code generation might have an option for even this; have a look here: https://github.com/filecoin-project/mir/tree/main/codegen#mir-structs

File: pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (matejpavlovic) [reply pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go:150]: I see... From the way the function is used, it would make sense to me to have the `sendIfReady` function return a `bool` saying whether a new cert should be created (or maybe even trigger it itself, but not sure about that). What do you think?

File: pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (matejpavlovic) [reply pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go:183]: That is not quite the case, unless we use follow-up events in some smart way. We must have a confirmation that the data is stored on disk before sending the message.

Feel free to resolve this one after reading.

File: pkg/availability/multisigcollector/internal/parts/certverification/certverification.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (matejpavlovic) [reply pkg/availability/multisigcollector/internal/parts/certverification/certverification.go:163]: What I meant was changing the variable name, not error text.

File: pkg/availability/multisigcollector/internal/parts/certverification/certverification.go (hunk 3 above)
Comment (matejpavlovic) [reply pkg/availability/multisigcollector/internal/parts/certverification/certverification.go:111]: Then I suggest changing this condition to not check again here.

**Comment by matejpavlovic** [comment]
Thanks! I think one more pass is needed (see unresolved conversations), then I can take over.

File: pkg/availability/multisigcollector/internal/common/common.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/common/common.go]: done

File: pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go]: `NodeID` and `SeqNr` are named types, not declared as typed aliases of an unnamed type (like is the case for `BatchID`. See previous comment for an explanation of why `BatchID` is a type alias.

File: pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go]: I am already using the generated code derived from declaring these proto messages are mir structs, but that generated code does not contain a constructor function.

File: pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go:224]: ditto (2)

File: pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go:244]: ditto (2)

File: pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certcreation/certcreation.go:150]: `sendIfReady` should then check if that particular certificate is ready to be sent, and not if *any* is (as is the case now) in order to return the bool. 

If this was Rust, then keeping a reference (pointer) to the value should be enough, but in Go keeping a pointer to a value of a map is undefined. I think the most correct and easiest conceptually at the moment is to leave it as is.

File: pkg/availability/multisigcollector/internal/parts/certverification/certverification.go (hunk 3 above)
Comment (ranchalp) [reply pkg/availability/multisigcollector/internal/parts/certverification/certverification.go:111]: done

**Comment by ranchalp** [comment]
Thanks! done

File: pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (matejpavlovic) [reply pkg/availability/multisigcollector/internal/parts/batchreconstruction/batchreconstruction.go]: Ah ok I see, I first didn't properly understand the difference between named types and type aliases. Fair enough, resolving this convo.

**Review by matejpavlovic — APPROVED** [review]
Great job!
I'll take care of the rebasing and then we can merge.

**Comment by matejpavlovic** [comment]
This PR continues as #357 (squashed and rebased on top of main).
